### ISSUE TITLE
SoundEffectInstance reuse causing InstancePlayLimitException

### ISSUE THREAD

**Issue by HarriOrri** [MonoGame/MonoGame#7648]
I've been having issues when I stop and start a SoundEffectInstance quicker than the length of that sound. The following code snippet is sometimes called repeatedly for a while every 0.5s while the sound is 1s long:

`
AttackSound?.Stop();
AttackSound?.Play();
`

Essentially I want to cancel the previous SoundEffectInstance and stop it from playing, then restart it from the start. However, when I do this the SoundEffectInstancePool slowly fills up the list to 256 of that SoundEffectInstance, causing it to throw an InstancePlayLimitException at me. This only happens when the time between calls to the previous code snippet is shorter than the sound effect itself and it's called repeatedly for a while.

#### What version of MonoGame does the bug occur on:
- MonoGame 3.8

#### What operating system are you using:
- Windows and Mac

#### What MonoGame platform are you using:

- DesktopGL

**Comment by Vectovox** [comment]
I just encountered something similar (just googled it, which is why I am here). However, I only reuse the instance once it has stopped by itself.

```
if (_instance.State != SoundState.Playing)
{
    _instance.Play();
}
```

It took maybe 10 minutes for it to happen, and the sound effect is around 400ms long.

May I suggest tagging the issue with Audio as well? @HarriOrri

**Comment by HarriOrri** [comment]
Yeah, I noticed that this won't happen if I let the sound finish playing before I try to play it again. however, I want to "reset" the sound, so hard-stop it and start from the beginning.

Also, it seems like I can't add labels to issues. Seems like only the repo maintainers can.

**Comment by Vectovox** [comment]
> I noticed that this won't happen if I let the sound finish playing before I try to play it again

Interesting, because in my case it occurs even when I let it finish. Maybe it is not related but it feels related anyhow.

> Also, it seems like I can't add labels to issues. Seems like only the repo maintainers can.

Got it. My bad!

**Comment by Eversor** [comment]
Having the same issue on all platforms, Desktop, Android and IOS.

**Comment by mrhelmut** [comment]
Does anyone of you have a working test project to share?

As far as I see it, the information is incomplete because Stop=>Play alone doesn't expose this issue. However, I can clearly see a bug whenever Play is called twice without a Stop in-between (it doesn't recycle the pre-existing voice, which means they add up with the original one being lost in limbo, never to be freed).

So I can clearly see why @Vectovox's code fails, but I don't see why @HarriOrri's does.

Also, this isn't related to ```SoundEffectInstancePool``` because ```SoundEffectInstance``` created through ```SoundEffectInstance inst = _soundEffect.CreateInstance()``` aren't pooled (it's the purpose of ```CreateInstance()```, it gives the life cycle management back to you).
It is this line that you are hitting:
https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework/Platform/Audio/OpenALSoundController.cs#L419

Besides the Play=>Play bug, bear in mind that when using ```CreateInstance()```, you are responsible of calling ```Dispose()``` on the ```SoundEffectInstance``` whenever you don't have any use of it anymore. If you let it go to the garbage without disposing it first, there is a chance that you hit the voice limit before the garbage collector does its job.

**Comment by Eversor** [comment]
@mrhelmut i can confirm it happens to me if i reuse the SoundEffectInstance before it has finished playing and without calling Stop in between.

**Comment by mrhelmut** [comment]
I'll come up with a fix for 3.8.1 for the Play() => Play() scenario, it should be simple enough.

**Comment by HarriOrri** [comment]
> Does anyone of you have a working test project to share?
> 
> As far as I see it, the information is incomplete because Stop=>Play alone doesn't expose this issue. However, I can clearly see a bug whenever Play is called twice without a Stop in-between (it doesn't recycle the pre-existing voice, which means they add up with the original one being lost in limbo, never to be freed). Also, from the look of it, it's about DesktopGL, Android, and iOS. But WindowsDX and UWP shouldn't have this issue.
> 
> So I can clearly see why @Vectovox's code fails, but I don't see why @HarriOrri's does.
> 
> Also, this isn't related to `SoundEffectInstancePool` because `SoundEffectInstance` created through `SoundEffectInstance inst = _soundEffect.CreateInstance()` aren't pooled (it's the purpose of `CreateInstance()`, it gives the life cycle management back to you). It is this line that you are hitting:
> 
> https://github.com/MonoGame/MonoGame/blob/64d82a10d32821a670e8040ca9aecf4e0024997d/MonoGame.Framework/Platform/Audio/OpenALSoundController.cs#L419
> 
> Besides the Play=>Play bug, bear in mind that when using `CreateInstance()`, you are responsible of calling `Dispose()` on the `SoundEffectInstance` whenever you don't have any use of it anymore. If you let it go to the garbage without disposing it first, there is a chance that you hit the voice limit before the garbage collector does its job.

I created a test project which showcases the problem we encountered. [Here's a link to it](https://github.com/surprised-turtle/InstancePlayLimitException-TestProject). It throws the exception here:

https://github.com/MonoGame/MonoGame/blob/f3a34a5f1850ad2adf3c4e47087bdeab55280d57/MonoGame.Framework/Audio/SoundEffectInstance.cs#L160

Our sound has a duration of 1.049 seconds but we replay it every roughly 0.74 using the following method of stopping it and playing it again

```
            _soundDelay -= gameTime.ElapsedGameTime.TotalSeconds;

            if (_soundDelay < 0.0)
            {
                _badSoundInstance.Stop();
                _badSoundInstance.Play();
                _soundDelay = 22.0 / 30.0;
            }
```

This will slowly fill up the `SoundEffectInstancePool._playingInstances` with the same SoundEffectInstance one by one every time we run the above code snippet, until it finally throws the `InstancePlayLimitException`

**Comment by mrhelmut** [comment]
Should be fixed by #7776

There was two different issues feeding each other.

**Comment by HarriOrri** [comment]
Awesome, thank you very much for your help!
